src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,501 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+        server& srv;
+        cluster::partition_manager& mgr;
+        ss::abort_source& as;
+    };
+
+    // Returned by fetch workers to the coordinator shard.
+    struct worker_result {
+        std::vector<read_result> read_results;
+        std::vector<size_t> errored_partitions;
+        size_t total_size;
+    };
+
+public:
+    nonpolling_fetch_plan_executor()
+      : _last_result_size(ss::smp::count, 0) {}
+
+    /**
+     * Executes the supplied `plan` until `octx.should_stop_fetch` returns true.
+     */
+    ss::future<> execute_plan(op_context& octx, fetch_plan plan) final {
+        co_await _as.start(octx.rctx.abort_source().local());
+
+        // start fetching from a random shard to make sure that we fetch data
+        // from all the partitions even if we reach fetch message size limit
+        const ss::shard_id start_shard_idx = random_generators::get_int(
+          ss::smp::count - 1);
+        for (size_t i = 0; i < ss::smp::count; ++i) {
+            auto shard = (start_shard_idx + i) % ss::smp::count;
+
+            ssx::spawn_with_gate(_workers_gate, [&]() mutable {
+                return log_and_ignore_exceptions(start_shard_fetch_worker(
+                  shard, octx, std::move(plan.fetches_per_shard[shard]), 0));
+            });
+        }
+
+        for (;;) {
+            co_await progress_conditions(octx);
+
+            if (octx.should_stop_fetch()) {
+                // Send abort signal to workers and wait for all workers to end
+                // here.
+                co_await _as.request_abort();
+                co_await _workers_gate.close();
+                co_await _as.stop();
+                co_return;
+            }
+
+            std::vector<shard_fetch> completed_shard_fetches = std::move(
+              _completed_shard_fetches);
+            _completed_shard_fetches.clear();
+
+            for (auto& sf : completed_shard_fetches) {
+                auto shard = sf.shard;
+                ssx::spawn_with_gate(
+                  _workers_gate,
+                  [this, &octx, shard, sf = std::move(sf)]() mutable {
+                      return log_and_ignore_exceptions(start_shard_fetch_worker(
+                        shard,
+                        octx,
+                        std::move(sf),
+                        // Require that a worker returns more data than before.
+                        // Otherwise it'll return right away with the previous
+                        // result.
+                        _last_result_size[shard] + 1));
+                  });
+            }
+        }
+    }
+
+private:
+    /**
+     * Determines if the should_stop_fetch() condition should be checked again.
+     * The return future is set if;
+     * - octx.deadline has been reached.
+     * - _as has been aborted.
+     * - one of the shard workers has returned results.
+     */
+    ss::future<> progress_conditions(op_context& octx) {
+        ss::promise<> timeout_pr, abort_pr;
```

Comment:
Could you just use the same promise for both conditions? Seems like it should be faster. Though I guess you have to check if it has already been set.

Reply:
AFAIK if a promise is set twice it'll throw. So I'd have to check first. And at that point it'd be better to just have two promises. Will double check though.